File: nucliadb_vectors/tests/test_hidden.rs
// Copyright (C) 2021 Bosutech XXI S.L.
//
// nucliadb is offered under the AGPL v3.0 and as commercial software.
// For commercial licensing, contact us at info@nuclia.com.
//
// AGPL:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//

use std::collections::{HashMap, HashSet};

use nucliadb_core::protos::{
    IndexParagraph, IndexParagraphs, Resource, ResourceId, VectorSearchRequest, VectorSentence,
};
use nucliadb_core::query_language::BooleanExpression;
use nucliadb_core::vectors::{MergeParameters, ResourceWrapper, VectorReader, VectorWriter, VectorsContext};
use nucliadb_core::NodeResult;
use nucliadb_vectors::config::VectorConfig;
use nucliadb_vectors::service::{VectorReaderService, VectorWriterService};
use tempfile::tempdir;
use uuid::Uuid;

fn resource(labels: Vec<String>) -> Resource {
    let id = Uuid::new_v4().to_string();
    Resource {
        resource: Some(ResourceId {
            shard_id: String::new(),
            uuid: id.clone(),
        }),
        labels,
        paragraphs: HashMap::from([(
            format!("{id}/a/title"),
            IndexParagraphs {
                paragraphs: HashMap::from([(
                    format!("{id}/a/title/0-5"),
                    IndexParagraph {
                        start: 0,
                        end: 5,
                        sentences: HashMap::from([(
                            format!("{id}/a/title/0-5"),
                            VectorSentence {
                                vector: vec![0.5, 0.5, 0.5, rand::random()],
                                metadata: None,
                            },
                        )]),
                        ..Default::default()
                    },
                )]),
            },
        )]),
        ..Default::default()
    }
}

#[test]
fn test_hidden_search() -> NodeResult<()> {
    let workdir = tempdir()?;
    let index_path = workdir.path().join("vectors");
    let mut writer = VectorWriterService::create(&index_path, String::new(), VectorConfig::default())?;

    // Create two resources, one hidden and one not
    let labels = vec!["/q/h".to_string()];
    let hidden_resource = resource(labels);
    writer.set_resource(ResourceWrapper::from(&hidden_resource))?;

    let visible_resource = resource(vec![]);
    writer.set_resource(ResourceWrapper::from(&visible_resource))?;

    // Find all resources
    let reader = VectorReaderService::open(&index_path)?;
    let request = VectorSearchRequest {
        vector: vec![0.5, 0.5, 0.5, 0.5],
        min_score: -1.0,
        result_per_page: 10,
        ..Default::default()
    };
    let all = reader.search(
        &request,
        &VectorsContext {
            filtering_formula: None,
            segment_filtering_formula: None,
        },
    )?;
    assert_eq!(
        HashSet::from_iter(all.documents.into_iter().map(|d| d.doc_id.unwrap().id)),
        HashSet::from([
            format!("{}/a/title/0-5", hidden_resource.resource.unwrap().uuid),
            format!("{}/a/title/0-5", visible_resource.resource.as_ref().unwrap().uuid)
        ])
    );

    // Find only the visible resource
    let visible = reader.search(
        &request,
        &VectorsContext {
            filtering_formula: None,
            segment_filtering_formula: Some(BooleanExpression::Not(Box::new(BooleanExpression::Literal(
                "/q/h".to_string(),
            )))),
        },
    )?;
    assert_eq!(visible.documents.len(), 1);
    assert_eq!(
        visible.documents[0].clone().doc_id.unwrap().id,
        format!("{}/a/title/0-5", visible_resource.resource.as_ref().unwrap().uuid)
    );

    // Won't merge results
    let merge_plan = writer.prepare_merge(MergeParameters {
        max_nodes_in_merge: 1000,
        segments_before_merge: 1,
        maximum_deleted_entries: 0,
    })?;
    assert!(merge_plan.is_none());

    Ok(())
}

File: nucliadb/tests/nucliadb/integration/search/test_hidden.py
# Copyright (C) 2021 Bosutech XXI S.L.
#
# nucliadb is offered under the AGPL v3.0 and as commercial software.
# For commercial licensing, contact us at info@nuclia.com.
#
# AGPL:
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
import asyncio

import pytest
from httpx import AsyncClient

from nucliadb.common.context import ApplicationContext
from nucliadb_protos.writer_pb2_grpc import WriterStub
from tests.utils import broker_resource_with_title_paragraph, inject_message


async def create_resource(kbid: str, nucliadb_ingest_grpc: WriterStub):
    message = broker_resource_with_title_paragraph(kbid)
    await inject_message(nucliadb_ingest_grpc, message)
    return message.uuid


@pytest.mark.deploy_modes("standalone")
async def test_hidden_search(
    app_context,
    nucliadb_reader: AsyncClient,
    nucliadb_writer: AsyncClient,
    nucliadb_ingest_grpc: WriterStub,
    nucliadb_writer_manager: AsyncClient,
    standalone_knowledgebox: str,
):
    resp = await nucliadb_writer_manager.patch(
        f"/kb/{standalone_knowledgebox}", json={"hidden_resources_enabled": True}
    )
    assert resp.status_code == 200

    r1 = await create_resource(standalone_knowledgebox, nucliadb_ingest_grpc)
    r2 = await create_resource(standalone_knowledgebox, nucliadb_ingest_grpc)

    # Both resources appear in searches
    resp = await nucliadb_reader.get(f"/kb/{standalone_knowledgebox}/search")
    assert resp.status_code == 200
    assert resp.json()["resources"].keys() == {r1, r2}

    resp = await nucliadb_reader.get(f"/kb/{standalone_knowledgebox}/suggest?query=title")
    assert resp.status_code == 200
    assert {r["rid"] for r in resp.json()["paragraphs"]["results"]} == {r1, r2}

    # Hide r1
    resp = await nucliadb_writer.patch(
        f"/kb/{standalone_knowledgebox}/resource/{r1}", json={"hidden": True}
    )
    assert resp.status_code == 200
    await asyncio.sleep(0.5)

    # Only r2 appears on search
    resp = await nucliadb_reader.get(f"/kb/{standalone_knowledgebox}/search")
    assert resp.status_code == 200
    assert resp.json()["resources"].keys() == {r2}

    resp = await nucliadb_reader.get(f"/kb/{standalone_knowledgebox}/suggest?query=title")
    assert resp.status_code == 200
    assert {r["rid"] for r in resp.json()["paragraphs"]["results"]} == {r2}

    # Unless show_hidden is passed, then both resources are returned
    resp = await nucliadb_reader.get(f"/kb/{standalone_knowledgebox}/search?show_hidden=true")
    assert resp.status_code == 200
    assert resp.json()["resources"].keys() == {r1, r2}

    resp = await nucliadb_reader.get(
        f"/kb/{standalone_knowledgebox}/suggest?query=title&show_hidden=true"
    )
    assert resp.status_code == 200
    assert {r["rid"] for r in resp.json()["paragraphs"]["results"]} == {r1, r2}

    # Test catalog ternary filter
    resp = await nucliadb_reader.get(f"/kb/{standalone_knowledgebox}/catalog")
    assert resp.status_code == 200
    assert resp.json()["resources"].keys() == {r1, r2}

    resp = await nucliadb_reader.get(f"/kb/{standalone_knowledgebox}/catalog?hidden=true")
    assert resp.status_code == 200
    assert resp.json()["resources"].keys() == {r1}

    resp = await nucliadb_reader.get(f"/kb/{standalone_knowledgebox}/catalog?hidden=false")
    assert resp.status_code == 200
    assert resp.json()["resources"].keys() == {r2}


@pytest.fixture()
async def app_context(natsd, storage, nucliadb):
    ctx = ApplicationContext()
    await ctx.initialize()
    yield ctx
    await ctx.finalize()
